mathics/builtin/procedural.py

```diff
@@ -472,11 +472,29 @@ def eval(self, n, evaluation: Evaluation):
             )
             return
 
+        # Due to the GIL lock, if we implement this method
+        # by just calling `time.sleep(sleeptime)`, the
+        # evaluation would not be aware of `TimeoutException`
+        # raised by an outer `TimeConstrained` expression.
+        #
+        # For this reason, the splitting of the sleep time is
+        # needed.
+        # It was also noticed in tests that in some platforms
+        # the total time that takes n calls to time.sleep(delta_t)
+        # can be appreciably larger than n*delta_t. For this reason,
+        # we also need to chech that inside the loop that the
+        # enlapsed time at the i-esim iteration does not exceed
+        # the  desired total time.
+
         steps = int(self.PAUSE_TICKS_PER_SECOND * sleeptime)
+        step_duration = 1 / self.PAUSE_TICKS_PER_SECOND
+        start = time.time()
```

Comment:
See https://stackoverflow.com/questions/8420422/python-windows-equivalent-of-sigalrm for a more robust mechanism that uses `Threading.Event` to allow Pause or sleep to be canceled.

Reply:
Involved code like this should be in eval.

Reply:
This should now go inside our `stopit` branch.

Reply:
I added a comment in mathics.eval.datetime about this.

Reply:
There is something about that comment that doesn't feel quite right, but I'd like to leave it until such time when we can devote resources to fully solving the problem.